### ISSUE TITLE
Use @Configuration annotation instead of kdoc in all rules

### ISSUE THREAD

**Issue by marschwar** [detekt/detekt#3670]
While the annotations `@ActiveByDefault` and `@RequiresTypeResolution` from #3562 have already been applied to all rules, the `@Configuration` introduced by #3637 has yet to be used in all rules. 
I propose to adapt the rules one module at a time and use this issue to keep track of the progress.  

- [x] detekt-rules-complexity (#3768 )
- [X] detekt-rules-coroutines (nothing to do)
- [X] detekt-rules-documentation
- [X] detekt-rules-empty
- [X] detekt-rules-errorprone
- [x] detekt-rules-exceptions (#3798 )
- [X] detekt-rules-naming
- [X] detekt-rules-performance (nothing to do)
- [x] detekt-rules-style (#3774 )
- [X] custom-checks
- [X] detekt-formatting
- [X] ~~replace annotations in ruleset providers~~ (#3816 )

**Once all rules have been adapted:**
- [x] Remove the checks in `ConfigAssert#checkOptions`
- [x] Remove the `ThresholdRule` ( #3891)
- [X] Remove support for `@configuration` kdoc tag in `io.gitlab.arturbosch.detekt.generator.collection.RuleVisitor` ~~and `RuleSetProviderVisitor`~~.
- [x] Remove `LazyRegex` ( #3891)

**Comment by marschwar** [comment]
I am working on `rules-complexity` ATM

**Comment by BraisGabin** [comment]
I was working on `detekt-rules-exceptions` and I had some problems with the rule `SwallowedException` (I had these problems in other rules too but this example illustrates the problem).

Current kdoc with the configuration:
```kotlin
/*
 * @configuration ignoredExceptionTypes - exception types which should be ignored (both in the catch clause and body)
 * (default: `- InterruptedException
 *            - NumberFormatException
 *            - ParseException
 *            - MalformedURLException`)
 * @configuration allowedExceptionNameRegex - ignores too generic exception types which match this regex
 * (default: `'_|(ignore|expected).*'`)
 */
```

and the properties are:
```kotlin
    private val ignoredExceptionTypes = valueOrDefaultCommaSeparated(IGNORED_EXCEPTION_TYPES, defaultIgnoredExceptions)
        .map { it.removePrefix("*").removeSuffix("*") }

    private val allowedExceptionNameRegex by LazyRegex(ALLOWED_EXCEPTION_NAME_REGEX, ALLOWED_EXCEPTION_NAME)
```

The problem here is that if I use `by config(defaultIgnoredExceptions)` I can't do `.map { it.removePrefix("*").removeSuffix("*") }`. I will need an extra property. How should we fix this? expand the `config()` to accept a lambda an do the changes there?

Something similar happens with `allowedExceptionNameRegex` and `LazyRegex`.

I'm thinking that, probably, with the lambda idea we don't even need to implement #3672.

**Comment by marschwar** [comment]
I had a similar issue in `LabeledExpression` I think. I did add an extra property such as:

**before**
```kotlin
    private val ignoredLabels = valueOrDefaultCommaSeparated(IGNORED_LABELS, emptyList())
        .map { it.removePrefix("*").removeSuffix("*") }
```

**after**

```kotlin
    @Configuration("allows to provide a list of label names which should be ignored by this rule")
    private val ignoredLabels: List<String> by config(emptyList())

    private val ignoredLabelsWithoutWildcards = ignoredLabels.map { it.removeSurrounding("*") }
```

Adding a lambda to the delegate would be possible but it would make it much harder to determine the default value (without type resolution). If possible I would prefer to keep the config delegate as slim as possible.

**Comment by chao2zhang** [comment]
Claiming `detekt-rules-errorprone`

**Comment by BraisGabin** [comment]
Yes, that's what I mean. Adding that type of lambda. It should not make the detection of the default value more difficult. The default value is the same. But we transform it as we want. Probably we should use some kind of lazy here to avoid multiple reading to the config, right?

**Comment by chao2zhang** [comment]
@BraisGabin I have some difficulty trying to understand the necessity of those custom transformation: After reading from yaml as `List<String>`, do we want to remove "*" from the prefix or suffix for each item in the list?

**Comment by BraisGabin** [comment]
Yes, in that case we want to do that. But my idea is to make it work for any case. For example in `TooGenericExceptionThrown` we want to change the `List<String>` to a `HashSet<String>` I imagine that we want it for performance reasons. The API would be somthing like:

```kotlin
@Configuration("exceptions which are too generic and should not be thrown")
private val exceptions: Set<String> by config(thrownExceptionDefaults) { it.toHashSet() }
```

or 

```kotlin
@Configuration("exception types which should be ignored (both in the catch clause and body)")
private val ignoredExceptionTypes: List<String> by config(defaultIgnoredExceptions) { value ->
  value.map { it.removePrefix("*").removeSuffix("*") } 
}
```

or

```kotlin
@Configuration("ignores too generic exception types which match this regex")
private val allowedExceptionNameRegex: Regex by config(ALLOWED_EXCEPTION_NAME) { it.toRegex() }
```

**Comment by BraisGabin** [comment]
And now that I'm reading these sample I thinkt that we should hardcode the default values inside  `config()`. Those constants only make the code less readable.

**Comment by zmunm** [comment]
It seems that `:custom-checks` should also be added to the list. I will take it.

**Comment by marschwar** [comment]
Still waiting on #3675 and #3676 but will continue with `detekt-rules-documentation` next.

**Comment by BraisGabin** [comment]
Yes, I vote to merge them. We can iterate later if we find more issues.

**Comment by marschwar** [comment]
> I was working on detekt-rules-exceptions 

Are you working on this at the moment @BraisGabin? I could to pick up where you left of if you like.

**Comment by BraisGabin** [comment]
You can go. I did nearly nothing because I found the regex problem

**Comment by chao2zhang** [comment]
The `detekt-formatting` module probably needs migration as well (I can work on it).

**Comment by marschwar** [comment]
Well, it is done. Thank you everyone.

**Comment by chao2zhang** [comment]
Note that we still have two configurations in RuleSetProvider [FormattingProvider](https://github.com/detekt/detekt/blob/main/detekt-formatting/src/main/kotlin/io/gitlab/arturbosch/detekt/formatting/FormattingProvider.kt). This would be a separate task and I can take a look this week.

**Comment by marschwar** [comment]
That would be awesome. I could not come up with a solution so far. Please also check out #3816.

**Comment by chao2zhang** [comment]
As #3891 is merged, could we close this issue?
